### PR TITLE
Validate coinbase expiration height

### PR DESCRIPTION
**Pull request by oxarbitrage**
## Motivation

We need to validate coinbase expiration height for nu5 and above. We also need to validate one rule related to expiration height for non coinbase transactions however this is not included in this PR.

This is part of https://github.com/ZcashFoundation/zebra/issues/2387 but it is not closing it, for that at least one rule more to non coinbase transactions need to be applied.

### Specifications

https://zips.z.cash/protocol/protocol.pdf#txnencodingandconsensus
https://zips.z.cash/zip-0203

## Solution

In this PR i am adding test vectors for nu5, for the testnet. I use this test vectors for a simple case where the rules pass. WE might need some tests for the rule failing, unsure if that should be a must.

Then, 2 consensus rules for the coinbase transactions are implemented:

> from NU5 activation, the nExpiryHeight field of a coinbase transaction MUST be set equal to the block height.

and 

> Consensus rule: [Overwinter to Canopy inclusive, pre-NU5] nExpiryHeight MUST be less than or equal to 499999999.

To complete https://github.com/ZcashFoundation/zebra/issues/2387 we need also:

> [NU5 onward] nExpiryHeight MUST be less than or equal to 499999999 for non-coinbase transactions.

## Review

I think @conradoplg can take a look initially into this as he was working on related stuff, then @teor2345 before any serious merge attempt.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

## Follow Up Work

Validate the missing rule to close the ticket.

### CODE REVIEW

**Comment by oxarbitrage** [comment]
There are a few tests that seems to use fake generated transactions that are failing. We probably need to make some tweaks there, will fix on Monday.

File: zebra-consensus/src/block/check.rs [hunk 1]
```diff
@@ -248,3 +248,41 @@ pub fn merkle_root_validity(
...
+                match coinbase.expiry_height() {
+                    None => Err(TransactionError::CoinbaseExpiration)?,
+                    Some(expiry) => {
+                        if expiry != *height {
+                            return Err(TransactionError::CoinbaseExpiration)?;
+                        }
+                    }
+                }
+            }
+            // Consensus rule: [Overwinter to Canopy inclusive, pre-NU5] nExpiryHeight
```
Comment (teor2345) [inline zebra-consensus/src/block/check.rs:279]: Possible Bug: If there is no NU5 activation height, then the pre-NU5 rules should be checked. But this code skips the rule when there is no NU5 activation height.

This is how `zcashd` checks this rule when NU5 is not active:
https://github.com/zcash/zcash/blob/c7089d2086c971c59d65b24ab39604da6182f44d/src/main.cpp#L1158-L1166

File: zebra-chain/src/block/height.rs [hunk 2]
```diff
@@ -48,6 +48,14 @@ impl Height {
...
     /// alias instead.
     pub const MAX_AS_U32: u32 = Self::MAX.0;
+
+    /// The maximum expiration Height that is allowed in coinbase transactions
+    /// previous to Nu5 and in non-coinbase transactions from Nu5 activation height
+    /// and above.
+    ///
+    /// TODO: This is currently the same as `Height::MAX` but that change in #1113.
+    /// Remove this TODO when that happens.
+    pub const MAX_COINBASE_EXPIRATION_PRE_NU5: Height = Height(499_999_999);
```
Comment (teor2345) [inline zebra-chain/src/block/height.rs]: Confusing Name: The name says that the limit only applies to pre-NU5 coinbase transactions, but it also applies to non-coinbase transactions.

We usually name constants after the general rule, and then document any exceptions in the comments.

```suggestion
    pub const MAX_EXPIRY_HEIGHT: Height = Height(499_999_999);
```

File: zebra-chain/src/block/height.rs [hunk 3]
```diff
@@ -48,6 +48,14 @@ impl Height {
     /// `Height::MAX.0` can't be used in match range patterns, use this
     /// alias instead.
     pub const MAX_AS_U32: u32 = Self::MAX.0;
+
+    /// The maximum expiration Height that is allowed in coinbase transactions
```
Comment (teor2345) [inline zebra-chain/src/block/height.rs]: Confusing Documentation: The consensus rules are:

> [Overwinter to Canopy inclusive, pre-NU5] nExpiryHeight MUST be less than or equal to 499999999.
> 
> [NU5 onward] nExpiryHeight MUST be less than or equal to 499999999 for non-coinbase transactions.

Which means:
```suggestion
    /// The maximum expiration Height that is allowed in all transactions
```

File: zebra-chain/src/block/height.rs [hunk 4]
```diff
@@ -48,6 +48,14 @@ impl Height {
...
+    /// previous to Nu5 and in non-coinbase transactions from Nu5 activation height
+    /// and above.
+    ///
+    /// TODO: This is currently the same as `Height::MAX` but that change in #1113.
```
Comment (teor2345) [inline zebra-chain/src/block/height.rs:56]: Good catch!

File: zebra-consensus/src/block.rs [hunk 5]
```diff
@@ -188,6 +188,9 @@ where
             tx::check::coinbase_outputs_are_decryptable(coinbase_tx, network, height)?;
             check::subsidy_is_valid(&block, network)?;
 
+            // Validate `nExpiryHeight` consensus rules
```
Comment (teor2345) [inline zebra-consensus/src/block.rs:191]: Note for next PR:

Just a reminder that the next PR needs to implement this rule for non-coinbase transactions:
> [Overwinter onward] If a transaction is not a coinbase transaction and its nExpiryHeight field is nonzero, then it MUST NOT be mined at a block height greater than its nExpiryHeight.

File: zebra-consensus/src/block/check.rs (hunk 1 above)
Comment (teor2345) [inline zebra-consensus/src/block/check.rs:279]: Note for next PR:

Just a reminder that this rule also applies to non-coinbase transactions:
> [Overwinter to Canopy inclusive, pre-NU5] nExpiryHeight MUST be less than or equal to 499999999.
> 
> [NU5 onward] nExpiryHeight MUST be less than or equal to 499999999 for non-coinbase transactions.

In the next PR, let's split this rule out into a separate function, so we don't have to duplicate the code.

File: zebra-test/src/vectors/block.rs [hunk 6]
```diff
@@ -220,6 +220,10 @@ lazy_static! {
             (1_116_000, BLOCK_TESTNET_1116000_BYTES.as_ref()),
             (1_116_001, BLOCK_TESTNET_1116001_BYTES.as_ref()),
             (1_326_100, BLOCK_TESTNET_1326100_BYTES.as_ref()),
+            // Nu5
+            (1_599_199, BLOCK_TESTNET_1599199_BYTES.as_ref()),
+            (1_599_200, BLOCK_TESTNET_1599200_BYTES.as_ref()),
```
Comment (teor2345) [inline zebra-test/src/vectors/block.rs]: Can we move this comment?

Block `1_599_199` is the last Testnet Canopy block, but the comment makes it look like the first NU5 block.

```suggestion
            (1_599_199, BLOCK_TESTNET_1599199_BYTES.as_ref()),
            // Nu5
            (1_599_200, BLOCK_TESTNET_1599200_BYTES.as_ref()),
```

**Review by teor2345 — COMMENTED** [review]
Looks like a good start, but some rules still need to be checked when there is no NU5 activation height.

**Review by teor2345 — APPROVED** [review]
Looks good, let's merge!

**Pull request merged**